File: services/documents/src/documents/billing_client.py
"""S2S-klient mot billing (snapshot) och auth (tjänste-token).

Tjänste-token hämtas via OAuth2 client_credentials från auth och cachas i
Redis tills strax före utgång — samma mönster som
packages/shared/src/auth/service-token-client.ts på TS-sidan. TTL på
tjänste-token är 5 minuter (architecture.md #18), så cachen sparar en
runda per event utan att hålla ett dött token.

Snapshoten hämtas ur billings interna endpoint. X-Tenant-Id sätts från
eventets tenantId (architecture.md #17 — en autentiserad tjänst får hävda
tenant); X-Correlation-Id förs vidare (architecture.md #5).
"""

from __future__ import annotations

from typing import Any

import httpx
import redis.asyncio as redis_asyncio

from .config import Settings

_CACHE_PREFIX = "svc-token:"
_REFRESH_MARGIN_SECONDS = 30


class BillingClient:
    def __init__(self, settings: Settings, redis: redis_asyncio.Redis) -> None:
        self._settings = settings
        self._redis = redis
        self._scope = " ".join(sorted(settings.documents_client_scopes))
        self._cache_key = f"{_CACHE_PREFIX}{settings.documents_client_id}:{self._scope}"

    async def _service_token(self) -> str:
        cached = await self._redis.get(self._cache_key)
        if cached:
            return cached.decode("utf-8") if isinstance(cached, bytes) else cached
        return await self._refresh_token()

    async def _refresh_token(self) -> str:
        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.post(
                f"{self._settings.auth_base_url}/auth/token",
                json={
                    "grant_type": "client_credentials",
                    "client_id": self._settings.documents_client_id,
                    "client_secret": self._settings.documents_client_secret,
                    "scope": self._scope,
                },
            )
        if res.status_code != 200:
            raise RuntimeError(f"kunde inte hämta tjänste-token: {res.status_code}")
        body = res.json()
        token: str = body["access_token"]
        ttl = max(int(body.get("expires_in", 300)) - _REFRESH_MARGIN_SECONDS, 1)
        await self._redis.set(self._cache_key, token, ex=ttl)
        return token

    async def _invalidate_token(self) -> None:
        await self._redis.delete(self._cache_key)

    async def fetch_snapshot(
        self, *, invoice_id: int, tenant_id: int, correlation_id: str
    ) -> dict[str, Any]:
        token = await self._service_token()
        res = await self._get_snapshot(token, invoice_id, tenant_id, correlation_id)

        if res.status_code == 401:
            # architecture.md #19: 401 = "vem är du?" — ogiltigt eller
            # UTGÅNGET token. Precis det transienta felet som läker av sig
            # självt: cachen kan ha hunnit bli inaktuell mot billings
            # klocka, eller JWT_SERVICE_SECRET roterats. Kasta bort det
            # cachade tokenet och försök EN gång till med ett färskt —
            # lyckas inte det heller bubblar felet upp som transient
            # (konsumenten nack:ar/requeue:ar, i stället för att ack:a
            # bort en faktura som aldrig får sin PDF).
            await self._invalidate_token()
            token = await self._refresh_token()
            res = await self._get_snapshot(token, invoice_id, tenant_id, correlation_id)
            if res.status_code == 401:
                raise RuntimeError(
                    f"snapshot {invoice_id}: billing svarade 401 även efter nytt token"
                )

        if res.status_code == 404:
            raise SnapshotNotFound(invoice_id)
        if res.status_code == 403:
            # architecture.md #19: 403 = giltig token men FEL SCOPE (eller
            # en avstängd/borttagen tenant, se requireService i billing).
            # Ingen av dem löser sig av att vänta — permanent för
            # konsumenten (ack + larm), till skillnad från 401 ovan.
            raise SnapshotAccessDenied(invoice_id, res.status_code)
        if res.status_code != 200:
            raise RuntimeError(f"snapshot {invoice_id}: billing svarade {res.status_code}")
        return res.json()

    async def _get_snapshot(
        self, token: str, invoice_id: int, tenant_id: int, correlation_id: str
    ) -> httpx.Response:
        async with httpx.AsyncClient(timeout=15.0) as client:
            return await client.get(
                f"{self._settings.billing_base_url}/internal/invoices/{invoice_id}/snapshot",
                headers={
                    "authorization": f"Bearer {token}",
                    "x-tenant-id": str(tenant_id),
                    "x-correlation-id": correlation_id,
                },
            )


class SnapshotNotFound(RuntimeError):
    def __init__(self, invoice_id: int) -> None:
        self.invoice_id = invoice_id
        super().__init__(f"ingen snapshot för faktura {invoice_id}")


class SnapshotAccessDenied(RuntimeError):
    def __init__(self, invoice_id: int, status_code: int) -> None:
        self.invoice_id = invoice_id
        self.status_code = status_code
        super().__init__(f"snapshot {invoice_id}: billing nekade åtkomst ({status_code})")

File: services/payments/src/billing-client.ts
// S2S-klient mot billing (bankgiro->tenant, OCR/id->aktuell faktura) och
// auth (tjänste-token). Tjänste-token hämtas via OAuth2 client_credentials
// och cachas i Redis tills strax före utgång — getServiceToken i
// @faktura/shared, samma mönster som documents billing_client.py
// (kommentaren där syftar redan på den här TS-implementationen).
//
// X-Tenant-Id sätts från den tenant matchningsmotorn själv slagit upp
// (architecture.md #17 — en autentiserad tjänst får hävda tenant) — utom
// på resolveTenantByBankgiro, där det ÄR det anropet som ska avgöra
// tenanten.

import { getServiceToken } from "@faktura/shared";
import type Redis from "ioredis";
import { config } from "./config";

export interface InvoiceResolution {
  currentInvoiceId: number;
  customerId: number;
  status: string;
  currency: string;
  totalInclVatOre: number;
  paidOre: number;
  remainingOre: number;
}

export class BillingAccessDeniedError extends Error {
  constructor(public readonly statusCode: number) {
    super(`billing nekade åtkomst (${statusCode})`);
    this.name = "BillingAccessDeniedError";
  }
}

export class BillingClient {
  constructor(private readonly redis: Redis) {}

  private async serviceToken(forceFresh: boolean): Promise<string> {
    if (forceFresh) {
      // getServiceToken cachar på (clientId, scope) — samma nyckel som
      // en normal hämtning. Ett explicit cache-bortkast görs inte här:
      // en 401 följt av en ny getServiceToken-runda inom samma sekund
      // skulle ändå bara returnera samma (nu ogiltiga) cachade värde.
      // I stället låter vi Redis-TTL:n (satt utifrån tokenets egen
      // expires_in) vara sanningen — se retry-kommentaren i get().
      await this.redis.del(
        `svc-token:${config.paymentsClientId}:${[...config.paymentsClientScopes].sort().join(" ")}`,
      );
    }
    return getServiceToken({
      authBaseUrl: config.authBaseUrl,
      clientId: config.paymentsClientId,
      clientSecret: config.paymentsClientSecret,
      redis: this.redis,
      scopes: config.paymentsClientScopes,
    });
  }

  private async get(
    path: string,
    headers: Record<string, string>,
  ): Promise<{ status: number; body: unknown }> {
    const token = await this.serviceToken(false);
    let res = await fetch(`${config.billingBaseUrl}${path}`, {
      headers: { authorization: `Bearer ${token}`, ...headers },
    });
    if (res.status === 401) {
      // architecture.md #19: 401 = ogiltigt/utgånget token — transient,
      // läker av ett enda nytt försök med ett färskt token (samma mönster
      // som documents billing_client.py).
      const fresh = await this.serviceToken(true);
      res = await fetch(`${config.billingBaseUrl}${path}`, {
        headers: { authorization: `Bearer ${fresh}`, ...headers },
      });
    }
    const body = res.status === 204 ? undefined : await res.json().catch(() => undefined);
    return { status: res.status, body };
  }

  /**
   * Bankgiro -> tenant. undefined om inget bankgiro matchar (404) —
   * matchningsmotorn tolkar det som unknown_bankgiro, inte som ett fel.
   */
  async resolveTenantByBankgiro(
    bankgiro: string,
    correlationId: string,
  ): Promise<number | undefined> {
    const { status, body } = await this.get(
      `/internal/company-settings/by-bankgiro?bankgiro=${encodeURIComponent(bankgiro)}`,
      { "x-correlation-id": correlationId },
    );
    if (status === 404) return undefined;
    if (status === 403) throw new BillingAccessDeniedError(status);
    if (status !== 200) throw new Error(`by-bankgiro: billing svarade ${status}`);
    return (body as { tenantId: number }).tenantId;
  }

  /**
   * OCR -> aktuell faktura (kedjeföljning redan gjord av billing).
   * undefined om OCR:et inte matchar någon faktura för tenanten (404).
   */
  async resolveInvoiceByOcr(
    tenantId: number,
    ocr: string,
    correlationId: string,
  ): Promise<(InvoiceResolution & { matchedInvoiceId: number }) | undefined> {
    const { status, body } = await this.get(
      `/internal/invoices/by-ocr?ocr=${encodeURIComponent(ocr)}`,
      {
        "x-tenant-id": String(tenantId),
        "x-correlation-id": correlationId,
      },
    );
    if (status === 404) return undefined;
    if (status === 403) throw new BillingAccessDeniedError(status);
    if (status !== 200) throw new Error(`by-ocr: billing svarade ${status}`);
    return body as InvoiceResolution & { matchedInvoiceId: number };
  }

  /**
   * id -> aktuell faktura, med samma kedjeföljning. Används av den
   * manuella matchningen (admin anger ett invoiceId, verifieras mot en
   * LEVANDE remainingOre — domain.md #27). undefined om fakturan inte
   * finns för tenanten (404, inklusive tenant-gränsen — architecture.md,
   * "404 över tenant-gränsen, aldrig 403").
   */
  async resolveInvoiceById(
    tenantId: number,
    invoiceId: number,
    correlationId: string,
  ): Promise<InvoiceResolution | undefined> {
    const { status, body } = await this.get(`/internal/invoices/${invoiceId}/current`, {
      "x-tenant-id": String(tenantId),
      "x-correlation-id": correlationId,
    });
    if (status === 404) return undefined;
    if (status === 403) throw new BillingAccessDeniedError(status);
    if (status !== 200) throw new Error(`invoices/:id/current: billing svarade ${status}`);
    return body as InvoiceResolution;
  }
}
